File: Flora.Shared/UuidV5.cs
using System.Security.Cryptography;
using System.Text;

namespace Flora.Shared;

/// <summary>RFC 4122 UUID v5 (SHA-1), совместимый с пакетом <c>uuid</c> v5 в Apps/Web (namespace + UTF-8 name).</summary>
public static class UuidV5
{
    /// <summary>DNS namespace из RFC; совпадает с <c>uuid.NAMESPACE_DNS</c> / FLORA_UUID_NAMESPACE в <c>lib/fscp/constants.ts</c>.</summary>
    public static readonly Guid FloraNamespaceDnsScope = Guid.Parse("6ba7b810-9dad-11d1-80b4-00c04fd430c8");

    public static Guid FromNamespaceAndUtf8Name(Guid namespaceId, string name) =>
        FromNamespaceAndUtf8Name(namespaceId, Encoding.UTF8.GetBytes(name));

    public static Guid FromNamespaceAndUtf8Name(Guid namespaceId, ReadOnlySpan<byte> nameUtf8)
    {
        var ns = NamespaceAsBigEndian(namespaceId);
        var buf = new byte[ns.Length + nameUtf8.Length];
        ns.CopyTo(buf);
        nameUtf8.CopyTo(buf.AsSpan(ns.Length));
        var hash = SHA1.HashData(buf);
        return GuidV5FromHash16(hash.AsSpan(0, 16));
    }

    /// <summary>Идентификатор DM 1:1 как в <c>dmConversationUuid</c> (TypeScript).</summary>
    public static Guid DmConversationUuid(Guid userA, Guid userB)
    {
        var a = userA.ToString("d");
        var b = userB.ToString("d");
        string x;
        string y;
        if (string.CompareOrdinal(a, b) <= 0)
        {
            x = a;
            y = b;
        }
        else
        {
            x = b;
            y = a;
        }

        return FromNamespaceAndUtf8Name(FloraNamespaceDnsScope, $"{x}|{y}|fscp-dm-v1");
    }

    /// <summary>Идентификатор agreement public key как в <c>agreementPublicKeyId</c> (TypeScript).</summary>
    public static Guid AgreementPublicKeyId(Guid userUuid, Guid keyEpochId)
    {
        var u = userUuid.ToString("d");
        var e = keyEpochId.ToString("d");
        return FromNamespaceAndUtf8Name(FloraNamespaceDnsScope, $"{u}|{e}|agreement-v1");
    }

    private static ReadOnlySpan<byte> NamespaceAsBigEndian(Guid namespaceId)
    {
        var b = namespaceId.ToByteArray();
        return new[]
        {
            b[3], b[2], b[1], b[0],
            b[5], b[4],
            b[7], b[6],
            b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15],
        };
    }

    private static Guid GuidV5FromHash16(ReadOnlySpan<byte> hash16)
    {
        Span<byte> h = stackalloc byte[16];
        hash16.CopyTo(h);
        h[6] = (byte)((h[6] & 0x0F) | (5 << 4));
        h[8] = (byte)((h[8] & 0x3F) | 0x80);
        return new Guid(new byte[]
        {
            h[3], h[2], h[1], h[0],
            h[5], h[4],
            h[7], h[6],
            h[8], h[9], h[10], h[11], h[12], h[13], h[14], h[15],
        });
    }
}

File: Backend/crates/flora-shared/src/uuid_v5.rs
//! RFC 4122 UUID v5 (SHA-1). Порт `Flora.Shared/UuidV5.cs`; байтовый паритет с ним
//! и с TS-реализацией (`lib/fscp/deriveIds.ts` в client-core) доказан golden-векторами.

use uuid::Uuid;

/// DNS namespace из RFC; совпадает с `uuid.NAMESPACE_DNS` / `FLORA_UUID_NAMESPACE` (TS)
/// и `UuidV5.FloraNamespaceDnsScope` (C#).
pub const FLORA_NAMESPACE_DNS_SCOPE: Uuid = uuid::uuid!("6ba7b810-9dad-11d1-80b4-00c04fd430c8");

/// Аналог `UuidV5.FromNamespaceAndUtf8Name` (namespace + UTF-8 name).
pub fn from_namespace_and_utf8_name(namespace: &Uuid, name: &str) -> Uuid {
    Uuid::new_v5(namespace, name.as_bytes())
}

/// Идентификатор DM 1:1 — `"{min}|{max}|fscp-dm-v1"`, ordinal-сортировка строковых UUID.
/// Паритет: `UuidV5.DmConversationUuid` (C#) и `dmConversationUuid` (TS).
pub fn dm_conversation_uuid(user_a: &Uuid, user_b: &Uuid) -> Uuid {
    // Uuid::to_string() — всегда lowercase с дефисами (как Guid.ToString("d") и TS toLowerCase),
    // сравнение &str в Rust побайтовое — эквивалент string.CompareOrdinal.
    let a = user_a.to_string();
    let b = user_b.to_string();
    let (x, y) = if a <= b { (a, b) } else { (b, a) };
    from_namespace_and_utf8_name(&FLORA_NAMESPACE_DNS_SCOPE, &format!("{x}|{y}|fscp-dm-v1"))
}

/// Идентификатор agreement public key — `"{user}|{epoch}|agreement-v1"`.
/// Паритет: `UuidV5.AgreementPublicKeyId` (C#) и `agreementPublicKeyId` (TS).
pub fn agreement_public_key_id(user_uuid: &Uuid, key_epoch_id: &Uuid) -> Uuid {
    from_namespace_and_utf8_name(
        &FLORA_NAMESPACE_DNS_SCOPE,
        &format!("{user_uuid}|{key_epoch_id}|agreement-v1"),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    // Полный паритет со значениями C#/TS проверяется на golden-векторах в Tests/parity;
    // здесь — инварианты самой реализации.

    #[test]
    fn dm_uuid_is_order_independent() {
        let a = uuid::uuid!("aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa");
        let b = uuid::uuid!("bbbbbbbb-bbbb-4bbb-8bbb-bbbbbbbbbbbb");
        assert_eq!(dm_conversation_uuid(&a, &b), dm_conversation_uuid(&b, &a));
    }

    #[test]
    fn v5_output_has_version_5_and_rfc_variant() {
        let id = from_namespace_and_utf8_name(&FLORA_NAMESPACE_DNS_SCOPE, "flora.example");
        assert_eq!(id.get_version_num(), 5);
        assert_eq!(id.get_variant(), uuid::Variant::RFC4122);
    }
}
